rfcs/rfc-0004-role-and-everyone-mentions.md

```diff
@@ -0,0 +1,88 @@
+```text
+<r:@01H2D2VBG4D2W4FTYM5G0T33WA>
+    ^ role id
+ ^ role type mention
+```
+and for everyone mentions it'd look like
+```text
+<*:@>
+```
+of course in the frontend it'd have to be similar to the already existing mention feature where along with users server roles appear visibly separated from the users. 
+
+As for Everyone Mentions they'd appear at the top of the autocomplete selection to stop accidental everyone's. 
+
+Additionally when mentioning a large number of people (100+) it will show a confirmation dialogue that will ask for confirmation before the mention is sent. 
+
+## Bulk Mention Handling
+
+When a user mentions a selection of people the back-end should first determine what users are affected by the mention, in the case of an everyone mention this is the same as getting all the users in the server however for a role based selection this gets more complicated, 
+
+a good way to handle this would be to just query the database and ask it to only return users that match the role with potentially a limit to be able to use indexes to go through the users if the role mention is for a lot of people.
+
+Then to notify the users the back-end adds the notification per-user to a notification queue, and when the user connects that notification queue is queried and the notification is sent. 
+
+If the user is offline the notification is deferred until they connect once more.
+
+Once the user either dismisses or reads the associated message the notification is marked as resolved and deleted.
+
+## Corner Cases
+- Users named either with a role name or everyone. This is not a problem due to the fact that we separate out role and everyone mentions with special syntax instead of the standard user mention syntax.
+- Users joining in the middle of a bulk mention. In this case the users do not get mentioned.
+# Drawbacks
+
+- annoying users with notifications they did not want to be apart of
+- being abused by power-moderators
+- would add parsing time as it'd require the back-end to parse it further and do a complex selection over the affected users.
+# Rationale and alternatives
+
+- This solutions works the best as it's simple and we already have frontend components for mentions, and back-end infrastructure delivering notifications dynamically. It only requires additional parsing of content and easy to add permissions to happen. 
+- Adding multiple mentions by selecting affected roles or users using a table of some kind, this approach could work however there would need to be large changes done to the messaging API to facilitate such a solution.
+- A bot that sends a couple messages mentioning all of the affected usernames, I did not consider this to be a good solution as it practically DDoSes our servers and if used frequently could fill up chats with useless messages.
+
+# Prior art
+
+Discord, Slack and telegram solve this issue similarly and it's worked for them so far.
```

Comment:
that's a valid concern a potential alteration I could do is remove @everyone altogether and just stick to role mentions

Reply:
although only you felt that way about them

Reply:
i sorta half disagree.

i believe `@everyone` pings can be quite useful for very small servers, especially servers made for groups of friends or coworkers and such. they can also be useful for emergencies, like server raids for example. while i can very much see the appeal in making it opt-in by default, there's occasionally a certain role that only an `@everyone` ping can fulfill. making these opt-in can occasionally just be adding additional hoops for people to jump through, which could be incredibly annoying.

i think it'd make more sense to have `@everyone` opt-out by default, *but only in small servers* (<50 people). perhaps this could also apply for channels where only a small number of people have read access to.

Reply:
I like that idea